### PR TITLE
crypto: Upstream v0.35.x improvements

### PR DESCRIPTION
**Pull request by Yawning**
This is my stab at #9186.  Per discussion in the issue, the order in which things were done is different from how the changes went into v0.35.x for ease of doing the backport.  Note that for the sake of expediency I opted to do this manually instead of via git.

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Comment by cmwaters** [comment]
Thanks @Yawning. Just to confirm - none of these changes will break downstream users?

**Comment by Yawning** [comment]
> Thanks @Yawning. Just to confirm - none of these changes will break downstream users?

This (like the 0.35.x version) will change sr25519 key derivation because the v0.34.x method of doing so is wacky.  I'll amend the changelog to document this fact.

File: p2p/conn/secret_connection.go [hunk 1]
```diff
@@ -14,8 +14,8 @@ import (
 	"time"
 
 	gogotypes "github.com/gogo/protobuf/types"
-	"github.com/gtank/merlin"
 	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/oasisprotocol/curve25519-voi/primitives/merlin"
```
Comment (marbar3778) [inline p2p/conn/secret_connection.go:18]: This is more performant or?

Comment (Yawning) [reply p2p/conn/secret_connection.go:18]: The code's nicer (IMO), more full featured and it removes 2 dependencies.  IIRC my strobe implementation is more performant for "sensible" message sizes, but the difference shouldn't be significant or noticeable for something like this.

**Comment by Yawning** [comment]
> Just to confirm - none of these changes will break downstream users?

Oh.  I forgot to mention this.  Switching to ZIP-215 is technically a breaking change as it is possible to produce signatures that will pass or fail verification depending on the nitpicky details things like ZIP-215 was written to address.

`voi` certainly supports "Behaving exactly like `crypto/ed25519`", and having it do so will be a one line change to this branch.  However while it still will be faster, batch verification under the hood will be done serially for correctness reasons.

Looking at #5632, it appears that some of the sub-projects migrated to using ZIP-215 (for v0.35.x), so they would have to undo the change, or ZIP-215 happens...

**Comment by zmanian** [comment]
ZIP 215 is technically not backwards compatible but practically speaking it is backwards compatible.

**Review by zmanian — APPROVED** [review]

**Review by cmwaters — APPROVED** [review]
Thanks for porting this change. Do you mind fixing the conflicts and updating the branch and then we can merge.

**Pull request merged**